File: pkg/runtime/runtimes/go/tests/validation_test.go
package main

import (
	"testing"

	"pulserpc-go-runtime/pulserpc"
)

func TestValidateString(t *testing.T) {
	if err := pulserpc.ValidateString("test"); err != nil {
		t.Errorf("Expected nil error for string, got %v", err)
	}

	if err := pulserpc.ValidateString(123); err == nil {
		t.Error("Expected error for non-string value")
	}
}

func TestValidateInt(t *testing.T) {
	if err := pulserpc.ValidateInt(123); err != nil {
		t.Errorf("Expected nil error for int, got %v", err)
	}

	// JSON numbers are decoded as float64, so we accept them
	if err := pulserpc.ValidateInt(123.0); err != nil {
		t.Errorf("Expected nil error for float64 representing int, got %v", err)
	}

	if err := pulserpc.ValidateInt("123"); err == nil {
		t.Error("Expected error for non-int value")
	}
}

func TestValidateFloat(t *testing.T) {
	if err := pulserpc.ValidateFloat(123.45); err != nil {
		t.Errorf("Expected nil error for float64, got %v", err)
	}

	if err := pulserpc.ValidateFloat(123); err != nil {
		t.Errorf("Expected nil error for int, got %v", err)
	}

	if err := pulserpc.ValidateFloat("123.45"); err == nil {
		t.Error("Expected error for non-float value")
	}
}

func TestValidateBool(t *testing.T) {
	if err := pulserpc.ValidateBool(true); err != nil {
		t.Errorf("Expected nil error for bool, got %v", err)
	}

	if err := pulserpc.ValidateBool(false); err != nil {
		t.Errorf("Expected nil error for bool, got %v", err)
	}

	if err := pulserpc.ValidateBool("true"); err == nil {
		t.Error("Expected error for non-bool value")
	}
}

func TestValidateArray(t *testing.T) {
	typeDef := map[string]interface{}{
		"builtIn": "string",
	}
	allStructs := pulserpc.StructMap{}
	allEnums := pulserpc.EnumMap{}

	elementValidator := func(v interface{}) error {
		return pulserpc.ValidateType(v, typeDef, allStructs, allEnums, false)
	}

	arr := []interface{}{"a", "b", "c"}
	if err := pulserpc.ValidateArray(arr, elementValidator); err != nil {
		t.Errorf("Expected nil error for valid array, got %v", err)
	}

	invalidArr := []interface{}{"a", 123, "c"}
	if err := pulserpc.ValidateArray(invalidArr, elementValidator); err == nil {
		t.Error("Expected error for invalid array element")
	}

	if err := pulserpc.ValidateArray("not an array", elementValidator); err == nil {
		t.Error("Expected error for non-array value")
	}
}

func TestValidateMap(t *testing.T) {
	typeDef := map[string]interface{}{
		"builtIn": "int",
	}
	allStructs := pulserpc.StructMap{}
	allEnums := pulserpc.EnumMap{}

	valueValidator := func(v interface{}) error {
		return pulserpc.ValidateType(v, typeDef, allStructs, allEnums, false)
	}

	m := map[string]interface{}{
		"key1": 1,
		"key2": 2,
	}
	if err := pulserpc.ValidateMap(m, valueValidator); err != nil {
		t.Errorf("Expected nil error for valid map, got %v", err)
	}

	invalidMap := map[string]interface{}{
		"key1": "not an int",
	}
	if err := pulserpc.ValidateMap(invalidMap, valueValidator); err == nil {
		t.Error("Expected error for invalid map value")
	}

	if err := pulserpc.ValidateMap("not a map", valueValidator); err == nil {
		t.Error("Expected error for non-map value")
	}
}

func TestValidateEnum(t *testing.T) {
	allowedValues := []string{"VALUE1", "VALUE2", "VALUE3"}

	if err := pulserpc.ValidateEnum("VALUE1", "TestEnum", allowedValues); err != nil {
		t.Errorf("Expected nil error for valid enum value, got %v", err)
	}

	if err := pulserpc.ValidateEnum("INVALID", "TestEnum", allowedValues); err == nil {
		t.Error("Expected error for invalid enum value")
	}

	if err := pulserpc.ValidateEnum(123, "TestEnum", allowedValues); err == nil {
		t.Error("Expected error for non-string enum value")
	}
}

func TestValidateStruct(t *testing.T) {
	allStructs := pulserpc.StructMap{
		"TestStruct": pulserpc.StructDef{
			"fields": []interface{}{
				map[string]interface{}{
					"name": "field1",
					"type": map[string]interface{}{"builtIn": "string"},
				},
				map[string]interface{}{
					"name":     "field2",
					"type":     map[string]interface{}{"builtIn": "int"},
					"optional": true,
				},
			},
		},
	}
	allEnums := pulserpc.EnumMap{}

	structDef := allStructs["TestStruct"]
	value := map[string]interface{}{
		"field1": "test",
		"field2": 123,
	}

	if err := pulserpc.ValidateStruct(value, "TestStruct", structDef, allStructs, allEnums); err != nil {
		t.Errorf("Expected nil error for valid struct, got %v", err)
	}

	// Missing required field
	invalidValue := map[string]interface{}{
		"field2": 123,
	}
	if err := pulserpc.ValidateStruct(invalidValue, "TestStruct", structDef, allStructs, allEnums); err == nil {
		t.Error("Expected error for missing required field")
	}

	// Invalid field type
	invalidTypeValue := map[string]interface{}{
		"field1": 123, // should be string
	}
	if err := pulserpc.ValidateStruct(invalidTypeValue, "TestStruct", structDef, allStructs, allEnums); err == nil {
		t.Error("Expected error for invalid field type")
	}
}

func TestValidateType(t *testing.T) {
	allStructs := pulserpc.StructMap{
		"TestStruct": pulserpc.StructDef{
			"fields": []interface{}{
				map[string]interface{}{
					"name": "field1",
					"type": map[string]interface{}{"builtIn": "string"},
				},
			},
		},
	}
	allEnums := pulserpc.EnumMap{
		"TestEnum": pulserpc.EnumDef{
			"values": []interface{}{
				map[string]interface{}{"name": "VALUE1"},
			},
		},
	}

	// Test built-in types
	stringType := map[string]interface{}{"builtIn": "string"}
	if err := pulserpc.ValidateType("test", stringType, allStructs, allEnums, false); err != nil {
		t.Errorf("Expected nil error for string, got %v", err)
	}

	// Test optional
	if err := pulserpc.ValidateType(nil, stringType, allStructs, allEnums, true); err != nil {
		t.Errorf("Expected nil error for optional nil, got %v", err)
	}

	// Test array
	arrayType := map[string]interface{}{
		"array": map[string]interface{}{"builtIn": "string"},
	}
	if err := pulserpc.ValidateType([]interface{}{"a", "b"}, arrayType, allStructs, allEnums, false); err != nil {
		t.Errorf("Expected nil error for array, got %v", err)
	}

	// Test map
	mapType := map[string]interface{}{
		"mapValue": map[string]interface{}{"builtIn": "int"},
	}
	if err := pulserpc.ValidateType(map[string]interface{}{"key": 1}, mapType, allStructs, allEnums, false); err != nil {
		t.Errorf("Expected nil error for map, got %v", err)
	}

	// Test struct
	structType := map[string]interface{}{
		"userDefined": "TestStruct",
	}
	structValue := map[string]interface{}{
		"field1": "test",
	}
	if err := pulserpc.ValidateType(structValue, structType, allStructs, allEnums, false); err != nil {
		t.Errorf("Expected nil error for struct, got %v", err)
	}

	// Test enum
	enumType := map[string]interface{}{
		"userDefined": "TestEnum",
	}
	if err := pulserpc.ValidateType("VALUE1", enumType, allStructs, allEnums, false); err != nil {
		t.Errorf("Expected nil error for enum, got %v", err)
	}
}

File: pkg/runtime/runtimes/java/tests/ValidationTest.java
import barrister2.*;
import org.junit.Test;
import org.junit.Assert;
import java.util.*;

public class ValidationTest {

    @Test
    public void testValidateString() {
        // Valid string
        Validation.validateString("hello");

        // Invalid - null
        try {
            Validation.validateString(null);
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("Expected string"));
        }

        // Invalid - integer
        try {
            Validation.validateString(123);
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("Expected string"));
        }
    }

    @Test
    public void testValidateInt() {
        // Valid int
        Validation.validateInt(42);

        // Invalid - string
        try {
            Validation.validateInt("not an int");
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("Expected int"));
        }
    }

    @Test
    public void testValidateBool() {
        // Valid boolean
        Validation.validateBool(true);
        Validation.validateBool(false);

        // Invalid - string
        try {
            Validation.validateBool("true");
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("Expected bool"));
        }
    }

    @Test
    public void testValidateArray() {
        List<String> stringArray = Arrays.asList("a", "b", "c");

        // Valid array of strings
        Validation.validateArray(stringArray, Validation::validateString);

        // Invalid - element validation fails
        List<Object> mixedArray = Arrays.asList("a", 123, "c");
        try {
            Validation.validateArray(mixedArray, Validation::validateString);
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("validation failed"));
        }
    }

    @Test
    public void testValidateMap() {
        Map<String, String> stringMap = new HashMap<>();
        stringMap.put("key1", "value1");
        stringMap.put("key2", "value2");

        // Valid map of strings
        Validation.validateMap(stringMap, Validation::validateString);

        // Invalid - value validation fails
        Map<String, Object> mixedMap = new HashMap<>();
        mixedMap.put("key1", "value1");
        mixedMap.put("key2", 123);
        try {
            Validation.validateMap(mixedMap, Validation::validateString);
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("validation failed"));
        }
    }

    @Test
    public void testValidateEnum() {
        List<String> allowedValues = Arrays.asList("RED", "GREEN", "BLUE");

        // Valid enum value
        Validation.validateEnum("RED", "Color", allowedValues);

        // Invalid - not in allowed values
        try {
            Validation.validateEnum("YELLOW", "Color", allowedValues);
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("Invalid value"));
        }

        // Invalid - not a string
        try {
            Validation.validateEnum(123, "Color", allowedValues);
            Assert.fail("Expected IllegalArgumentException");
        } catch (IllegalArgumentException e) {
            Assert.assertTrue(e.getMessage().contains("Expected string"));
        }
    }
}
